log/logwriter/writer.go

```diff
@@ -1,27 +1,149 @@
 package logwriter
 
 import (
+	"strings"
+	"sync"
+	"unicode"
+
 	"github.com/bitrise-io/bitrise/log"
+	"github.com/bitrise-io/bitrise/log/corelog"
 )
 
-// LogWriter ...
-type LogWriter struct {
+var ansiEscapeCodeToLevel = map[corelog.ANSIColorCode]corelog.Level{
+	corelog.RedCode:     corelog.ErrorLevel,
+	corelog.YellowCode:  corelog.WarnLevel,
+	corelog.BlueCode:    corelog.InfoLevel,
+	corelog.GreenCode:   corelog.DoneLevel,
+	corelog.MagentaCode: corelog.DebugLevel,
+}
+
+type LogLevelWriter struct {
+	mux    sync.Mutex
 	logger log.Logger
+
+	currentColor corelog.ANSIColorCode
+	currentLevel corelog.Level
+	currentChunk string
 }
 
-// NewLogWriter ...
-func NewLogWriter(logger log.Logger) LogWriter {
-	return LogWriter{
+// NewLogLevelWriter ...
+func NewLogLevelWriter(logger log.Logger) *LogLevelWriter {
+	return &LogLevelWriter{
 		logger: logger,
 	}
 }
 
-func (w LogWriter) Write(p []byte) (n int, err error) {
+// TODO: handle if currentChunk is too big
+func (w *LogLevelWriter) Write(p []byte) (n int, err error) {
```

Comment:
This is a meaty function with a lot of cases in it. I wonder if it could simplified in any way. 

One idea I have is that if its enough to look for a single colour and its reset code. Our steps log stuff line by line and every line has a closing reset code. In the old implementation we did not care of the embedded colours because those do not contribute to the log level and will be rendered by the website. 

I will try to write some rudimentary pseudo code:
```
func (w *LogLevelWriter) Write(p []byte) (n int, err error) {
	chunk := string(p)
	colorCodes := findEmbeddedColors(chunk)

	if len(colorCodes) > 0 {
		w.colorStack.add(colorCodes)
	}

	resetColorCount := numbeOfResetColors(chunk)
	currentColor := ""
	for i := 0; i < resetColorCount; i++ {
		currentColor = w.colorStack.pop()
	}

	w.currentChunk = w.currentChunk + chunk

	if len(w.colorStack) == 0 {
		w.logger.LogMessage(w.currentChunk, convertCurrentColorToLevel(currentColor))
		w.currentChunk = ""
		return len(p), nil
	}

	return len(p), nil
}
```
I am not yet sure that this logic is correct but what it tries to do is that it collects how many colors are active. Because we are only looking for the outermost one it simply pops the saved colors of the stack as we get the reset codes. Once the stack is empty we log the message based on the last color code the stack had.

Reply:
This PR took a slightly different direction: the change in this file was the primary functional change to fix the log-level parsing issue.
We decided to do some preliminary work before fixing the original issue.
The remaining file changes are still valid technical improvements, so the function change was reverted and this PR is about technical improvements now.